### PR TITLE
[cpu-offloading] Implement prefetching for cpu offloading pass

### PR DESCRIPTION
**Pull request by mlazos**
- configs.py: Renamed default_offload → budget_limited_offload memory policy. Removed
   cpu_offload_all. Added cpu_offload_defer_n_layers and cpu_offload_prefetch_n_layers
  config fields.                                                                       
  - cpu_offload.py (bulk of the change):                    
    - Removed _find_tensor_dep function                                                
    - Implemented region-based defer_offload_waits — physically moves forward          
  wait_tensor ops N regions later via anchor.append() to overlap D2H transfers with    
  compute                                                                              
    - Implemented layer-aware prefetch_reloads — moves backward reload ops N layers    
  earlier to overlap H2D transfers with compute. Counts by layer transitions (not raw  
  regions) to handle interleaved _NOT_IN_LAYERS gradient accumulation nodes in backward
    - Fixed budget logging to report actually selected tensor sizes                    
    - Added tag_all_offloadable_activations with CPU memory budget support             
  - graph_utils.py: Removed inductor_decomposition skip, updated error message for     
  renamed policy                                                                       
  - passes.py: Threads new config fields through to apply_cpu_offload_pass             
  - tests/: Added missing config fields to test mocks, updated integration test        
  configs, resolved merge conflicts    

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #3166

### CODE REVIEW

**Comment by mlazos** [comment]
@claude can you review this and ensure that the prefetching per layer does not miss any edge cases and that the tests do not mask any behavior

**Comment by mlazos** [comment]
@claude can you review to see if we are missing any overlap opportunities or edge cases?

File: torchtitan/experiments/graph_trainer/graph_utils.py [hunk 1]
```diff
@@ -549,6 +549,10 @@ def get_joint_custom_passes_from_config(
                 f"Available joint passes: {list(AVAILABLE_JOINT_PASSES.keys())}"
             )
 
+        # Skip inductor_decomposition - it's handled in joint_graph_builder
```
Comment (mlazos) [inline torchtitan/experiments/graph_trainer/graph_utils.py]: Not sure if this should be skipped

File: torchtitan/experiments/graph_trainer/passes.py [hunk 2]
```diff
@@ -824,6 +823,10 @@ def tag_with_memory_policy_pass(
         tag_all_offloadable_activations(gm)
     else:
         raise ValueError(f"Unknown memory_policy: {memory_policy!r}")
+
+    if config.compile.enable_cpu_offload and memory_policy != "cpu_offload_all":
```
Comment (mlazos) [inline torchtitan/experiments/graph_trainer/passes.py]: Review this

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 3]
```diff
@@ -343,42 +499,198 @@ def apply_cpu_offload_pass(
...
+        prefetch_n_layers: Move ao.reload nodes this many layers earlier in
+            the backward graph to overlap H2D with compute.
 
     Returns:
         The transformed GraphModule with offload/reload ops inserted.
     """
     tag_all_offloadable_activations(gm)
-    return apply_cpu_offload_pass(gm, example_inputs)
+
+    cpu_mem_budget = float(os.environ.get("OFFLOAD_CPU_MEMORY_BUDGET_GB", "100"))
```
Comment (mlazos) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: Maybe make a config for this

File: torchtitan/experiments/graph_trainer/configs.py [hunk 4]
```diff
@@ -67,6 +67,16 @@ class GraphTrainerCompileConfig(CompileConfig):
     enable_cudagraph: bool = True
     """When False, skip the cudagraph pass even if the graph is compatible."""
 
+    enable_cpu_offload: bool = False
```
Comment (mlazos) [inline torchtitan/experiments/graph_trainer/configs.py]: Hmm should probably remove

Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/configs.py]: yea, remove. "what to offload" is now controlled by memory_policy, and "how to offload" offload_pass is always applied.

File: torchtitan/experiments/graph_trainer/configs.py [hunk 5]
```diff
@@ -67,6 +67,16 @@ class GraphTrainerCompileConfig(CompileConfig):
...
+    """When True, tag saved activations for CPU offloading after the
+    memory_policy pass. Combines with SAC: memory_policy selects which
+    activations to save vs recompute, then offloading moves saved
+    activations to CPU."""
+
+    cpu_offload_prefetch_n_layers: int = 1
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/configs.py:75]: do we need this? 
Have you run into to a case where it need prefetch more then 1 layer?

File: torchtitan/experiments/graph_trainer/passes.py (hunk 2 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/passes.py]: don't need this. 
line at 822 is sufficient.

File: torchtitan/experiments/graph_trainer/passes.py [hunk 6]
```diff
@@ -169,10 +172,6 @@ def compile_time_passes(
 
     inductor_compilation = config.compile.inductor_compilation
     if inductor_compilation == "full":
-        # Compile the entire graph into optimized Triton kernels. Must
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/passes.py:178]: keep

File: torchtitan/experiments/graph_trainer/graph_utils.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/graph_utils.py]: don't need this here.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 3 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: yea, make it consistent.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 7]
```diff
@@ -343,42 +499,198 @@ def apply_cpu_offload_pass(
...
+        layer_first_wait[target_layer].prepend(reload_node)
+        moved += 1
+
+    if moved > 0:
+        logger.info(
+            f"CPU offload prefetch: moved {moved} reloads " f"{n_layers} layer(s) ahead"
+        )
+
+
 def cpu_offload_pass(
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: we can remove this pass now. 

tag_all_offloadable_activations is now run via tag_with_memory_policy_pass

File: torchtitan/experiments/graph_trainer/passes.py [hunk 8]
```diff
@@ -155,9 +156,11 @@ def compile_time_passes(
             tag_with_memory_policy_pass,
             config=config,
         ),
-        # TODO: currently either SAC or CPU offload is used, not both at the
-        # same time. Composability between these two passes is untested.
-        apply_cpu_offload_pass,
+        functools.partial(
+            apply_cpu_offload_pass,
+            prefetch_lookahead=config.compile.cpu_offload_prefetch_n_layers,
+            cpu_budget_gb=float(os.environ.get("OFFLOAD_CPU_MEMORY_BUDGET_GB", "100")),
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/passes.py]: cpu_budget_gb doesn't belong here. 

apply_cpu_offload_pass only take care of "how to offload" 
cpu_budget_gb is a limit for "what to offload", move it to tag_all_offloadable_activations

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 9]
```diff
@@ -229,17 +312,29 @@ def tag_all_offloadable_activations(
...
         if not _can_offload_node(node):
             continue
 
-        # Check for backward consumers
-        has_backward_users = any(_is_backward_node(u) for u in node.users)
-        if not has_backward_users:
-            continue
+        existing = node.meta.get("recompute")
+
+        if sac_active:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: don't need to detect sac_active? 
just do defensive programming, and assume some tagging has been applied

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 10]
```diff
@@ -343,42 +499,198 @@ def apply_cpu_offload_pass(
...
         f"CPU offload: offloaded {len(offloadable)} tensors "
-        f"({total_bytes / 1024 / 1024:.2f} MB)"
+        f"({total_bytes / 1024 / 1024:.2f} MB), "
+        f"repositioned {repositioned} forward waits"
     )
     return gm
 
 
+def _get_reload_layer(reload_node: Node) -> int:
+    """Get the layer ID of an offloaded activation from the forward op chain.
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: This is cumbersome, let's just save the module_fqn when creating the reload node, and read from there.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 11]
```diff
@@ -183,27 +258,34 @@ def _classify_forward_backward(
...
-      3. Passes ``_can_offload_node`` (not a view, not tiny, etc.)
-      4. Has at least one backward consumer
+    eligible forward nodes. When SAC is active, only MUST_SAVE nodes are
+    tagged (the surviving activations after recomputation decisions).
+    Without SAC, any forward node with backward consumers is eligible.
 
-    The last layer is skipped when multiple layers exist, since its activations
-    are consumed immediately in backward (offloading adds overhead with no
-    benefit).
+    The last layer is skipped (its activations are consumed immediately
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: mention that this is just and heuristic. 
In the case of chunked loss, it might still be beneficial to offload last layer.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 12]
```diff
@@ -335,6 +490,7 @@ def apply_cpu_offload_pass(
             )
             reload_node.meta["val"] = val
             reload_node.meta["autograd_backward"] = True
+            reload_node.meta["partitioner_tag"] = "must_be_in_backward"
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: no one is consuming partitioner_tag in graphtrainer.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 3 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: but we don't need cpu_mem_budget here.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 13]
```diff
@@ -343,42 +499,198 @@ def apply_cpu_offload_pass(
...
+        if layer_id != _NOT_IN_LAYERS:
+            reload_info.append((node, wait_node, layer_id))
+
+    if not reload_info:
+        return
+
+    node_to_idx: dict[Node, int] = {n: i for i, n in enumerate(gm.graph.nodes)}
+    reload_info.sort(key=lambda x: node_to_idx[x[1]])
+
+    layer_first_wait: dict[int, Node] = {}
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: any reason it need to be before the the first wait node? 
fsdp prefetch just place it at the beginning the layer, i.e. before first node in layer.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 14]
```diff
@@ -343,42 +499,198 @@ def apply_cpu_offload_pass(
...
+    offload = fwd_wait.args[0] if fwd_wait.args else None
+    if not isinstance(offload, Node):
+        return _NOT_IN_LAYERS
+    original = offload.args[0] if offload.args else None
+    if not isinstance(original, Node):
+        return _NOT_IN_LAYERS
+    return _get_layer_id(original)
+
+
+def prefetch_offloads(
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: prefetch_reloads

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 15]
```diff
@@ -343,42 +499,198 @@ def apply_cpu_offload_pass(
...
+    if not isinstance(fwd_wait, Node):
+        return _NOT_IN_LAYERS
+    offload = fwd_wait.args[0] if fwd_wait.args else None
+    if not isinstance(offload, Node):
+        return _NOT_IN_LAYERS
+    original = offload.args[0] if offload.args else None
+    if not isinstance(original, Node):
+        return _NOT_IN_LAYERS
+    return _get_layer_id(original)
+
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: I don't see defer_offload_waits?

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 16]
```diff
@@ -257,51 +352,110 @@ def tag_all_offloadable_activations(
...
+    GPU storage is freed inside ``ao.wait_tensor`` when ``keepalive`` is
+    provided: after the D2H sync completes, the keepalive tensor's storage
+    is released back to the allocator. The ``dep`` argument creates an
+    explicit scheduling dependency that prevents graph reordering passes
+    from moving the wait before the last forward consumer.
+
+    Forward waits are repositioned after the last forward consumer of each
+    GPU tensor's storage chain, maximizing D2H/compute overlap.
+
+    View replay: views share storage with the base tensor. We replay the
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: I am not sure if we want to support view relay atm, I think it's error-prone and non-trivial to get it right. 

An alternative is to just not offload views. Non-contiguous views are slow to offload anyway. 

I would prefer to do this in a different PR.

**Review by SherlockNoMad — CHANGES_REQUESTED** [review]

File: torchtitan/experiments/graph_trainer/configs.py (hunk 5 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/configs.py:75]: in my old experiments yes, but tbh in the new ones no, but I thought it best to leave it for experimentation at least.

File: torchtitan/experiments/graph_trainer/passes.py [hunk 17]
```diff
@@ -822,10 +823,16 @@ def tag_with_memory_policy_pass(
...
+            tag_all_offloadable_activations(
+                gm, cpu_budget_gb=config.compile.cpu_offload_budget_gb
+            )
     elif memory_policy == "eager":
         apply_sac_pass(gm, policy_fn=_make_eager_memory_policy())
     elif memory_policy == "cpu_offload_all":
-        tag_all_offloadable_activations(gm)
+        tag_all_offloadable_activations(
+            gm, cpu_budget_gb=config.compile.cpu_offload_budget_gb
+        )
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/passes.py]: let's deprecate this path. prefer default_offload

File: torchtitan/experiments/graph_trainer/configs.py [hunk 18]
```diff
@@ -44,14 +44,17 @@ class GraphTrainerCompileConfig(CompileConfig):
     debug_graph_passes: bool = False
     """Log timing, op-count diffs, and before/after graphs for each pass to tlparse."""
 
-    memory_policy: Literal["default", "eager", "cpu_offload_all"] = "default"
+    memory_policy: Literal[
+        "default", "eager", "default_offload", "cpu_offload_all"
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/configs.py]: remove cpu_offload_all

Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/configs.py]: also default_offload sounds like budget_limited_offload

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 19]
```diff
@@ -344,41 +446,160 @@ def apply_cpu_offload_pass(
...
+
+
+def defer_offload_waits(
     gm: torch.fx.GraphModule,
-    example_inputs: tuple | None = None,
-) -> torch.fx.GraphModule:
-    """Two-phase CPU offload pass: tag eligible activations, then insert ops.
+    wait_offload_map: dict[Node, Node],
+) -> int:
+    """Sink each forward wait_tensor to the next layer boundary.
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: unslop the wording: defer, sink, repositioned 
I prefer defer

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 20]
```diff
@@ -344,41 +446,160 @@ def apply_cpu_offload_pass(
...
+        node.meta["cpu_offload_reload_node"] = wait_node
+
         logger.debug(
             f"CPU offload: offloading {node.name} "
             f"({_tensor_bytes(val) / 1024:.1f} KB, {val.shape})"
         )
         total_bytes += _tensor_bytes(val)
 
+    # 4b. Defer forward waits to maximize D2H/compute overlap.
+    repositioned = defer_offload_waits(gm, wait_offload_map)
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: we have prefetch_lookahead, but no config for defer?

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 21]
```diff
@@ -257,31 +334,53 @@ def tag_all_offloadable_activations(
 # ============================================================
 
 
+def _find_tensor_dep(node: Node, node_positions: dict[Node, int]) -> Node | None:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: node_positions is not used??

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 22]
```diff
@@ -344,41 +446,160 @@ def apply_cpu_offload_pass(
...
+            if cl != _NOT_IN_LAYERS:
+                consumer_layer = max(consumer_layer, cl)
+
+        # Sink to next layer if possible, otherwise stay in consumer layer.
+        sink_layer = next_layer_map.get(consumer_layer, consumer_layer)
+        if sink_layer not in layer_last_fwd_node:
+            continue
+
+        anchor = layer_last_fwd_node[sink_layer]
+        dep_node = _find_tensor_dep(anchor, node_positions)
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: I am not sure why _find_tensor_dep is needed here? 

Since it's just for dep_node, anchor itself seems fine?

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 20 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: Added

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 22 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: Removed

File: torchtitan/experiments/graph_trainer/configs.py (hunk 18 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/configs.py]: changed to budget_limited_offload

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 23]
```diff
@@ -344,41 +434,166 @@ def apply_cpu_offload_pass(
...
+    for node in gm.graph.nodes:
+        if not (
+            node.op == "call_function"
+            and node.target == torch.ops.ao.reload.default
+            and _is_backward_node(node)
+        ):
+            continue
+
+        layer_id = _get_reload_layer(node)
+        if layer_id != _NOT_IN_LAYERS:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: I think harder about this... and we can do global optimization. 

emb, lmhead and loss are in  _NOT_IN_LAYERS

and the full graph is actually 
emb, layer0... layerN, lm_head , loss, loss_bw, (layer_bwN..0), emb_bw

we can prefetch/defer out of the layer.* boundary. 

So concretely, we can do it like this 
1. build an list of anchors globally, (include the non-layer region)
2. prefetch == move to an anchor before, defer == move to an anchor after.

Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: This is the unique kind of optimization you can do with full graph, with loss in the graph.

**Comment by mlazos** [comment]
@claude, can you look through any edge cases related to topological ordering

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 24]
```diff
@@ -335,6 +427,7 @@ def apply_cpu_offload_pass(
             )
             reload_node.meta["val"] = val
             reload_node.meta["autograd_backward"] = True
+            reload_node.meta["custom"] = dict(node_custom)
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: nit, node_custom is not used anywhere else. 
doesn't need to create node_custom

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 25]
```diff
@@ -344,41 +437,207 @@ def apply_cpu_offload_pass(
...
 
+    # 4b. Defer forward waits to maximize D2H/compute overlap.
+    if defer_n_layers > 0:
+        deferred = defer_offload_waits(gm, wait_offload_map, defer_n_layers)
+    else:
+        deferred = 0
+
+    # 5. Per-group backward reload prefetch.
+    if prefetch_lookahead > 0:
+        prefetch_reloads(gm, prefetch_lookahead)
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: prefetched = ...

for consistency?

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 26]
```diff
@@ -344,41 +437,207 @@ def apply_cpu_offload_pass(
...
+        if target_idx >= current_idx:
+            continue
+
+        bwd_anchors[target_idx].prepend(reload_node)
+        moved += 1
+
+    if moved > 0:
+        logger.info(
+            f"CPU offload prefetch: moved {moved} reloads {n_layers} layer(s) ahead"
+        )
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py:641]: return moved

**Review by SherlockNoMad — APPROVED** [review]
lgtm, let's land this and do some real runs. 

I would like to see some tlparse result for graph before and after the cpu_offload_passes. 

you can used https://www.internalfb.com/intern/paste/P2309998107/ to generate some tlparse.
